packages/flutter/lib/src/rendering/object.dart

```diff
@@ -6383,38 +6382,61 @@ final class _SemanticsGeometry {
         parentRenderObject = toParent;
       }
     }
-
-    // Calculate transform.
     assert(childToCommonAncestor.length >= 2);
-    for (int i = childToCommonAncestor.length - 1; i > 0; i -= 1) {
-      childToCommonAncestor[i].applyPaintTransform(childToCommonAncestor[i - 1], transform);
-    }
 
-    if (parentToCommonAncestorTransform != null) {
-      if (parentToCommonAncestorTransform.invert() != 0) {
-        transform.multiply(parentToCommonAncestorTransform);
-      } else {
-        transform.setZero();
-      }
-    }
+    // Calculate clips and transform.
 
-    // Calculate clips.
     Rect? paintClipRect;
     Rect? semanticsClipRect;
-    if (childToCommonAncestor.last == parent.renderObject) {
+    final Matrix4 transform;
+    if (parentToCommonAncestorTransform == null) {
       // This is most common case, i.e. parent is the common ancestor.
+      transform = Matrix4.identity();
       paintClipRect = parentPaintClipRect;
       semanticsClipRect = parentSemanticsClipRect;
-      assert(parentToCommonAncestorTransform == null);
+      // Travrese from `parent`'s render object to `child`'s.
       for (int i = childToCommonAncestor.length - 1; i > 0; i -= 1) {
-        (paintClipRect, semanticsClipRect) = _computeClipRect(
-          childToCommonAncestor[i],
-          childToCommonAncestor[i - 1],
-          semanticsClipRect,
-          paintClipRect,
+        final RenderObject nodeParent = childToCommonAncestor[i];
```

Comment:
For the more common codepath the transform is now computed in the same loop and there are less `inverse` calculations. The overlay portal path still uses `_computeClipRect` but that path usually isn't as hot.

Reply:
right, I don't think they can have this case where Render tree and semantics tree are differ without semanticsParent

Reply:
Sounds good, I don't think I've touched that code path in this patch so I'll probably replace that with an assert in a different PR.